src/librbd/CopyupRequest.cc

```diff
@@ -0,0 +1,150 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/ceph_context.h"
+#include "common/dout.h"
+#include "common/Mutex.h"
+
+#include "librbd/AioCompletion.h"
+#include "librbd/ImageCtx.h"
+
+#include "librbd/AioRequest.h"
+#include "librbd/CopyupRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::CopyupRequest: "
+
+namespace librbd {
+  CopyupRequest::CopyupRequest()
+    : m_ictx(NULL), m_object_no(0), m_lock("librbd::CopyupRequest::m_lock"),
+      m_ready(false), m_need_copyup(false), m_parent_completion(NULL),
+      m_copyup_completion(NULL) {}
+
+  CopyupRequest::CopyupRequest(ImageCtx *ictx, const std::string &oid,
+                               uint64_t objectno, bool need_copyup)
+    : m_ictx(ictx), m_oid(oid), m_object_no(objectno),
+      m_lock("librbd::CopyupRequest::m_lock"), m_ready(false),
+      m_need_copyup(need_copyup), m_parent_completion(NULL),
+      m_copyup_completion(NULL) {}
+
+  CopyupRequest::~CopyupRequest() {
+    assert(m_pending_requests.empty());
+
+    m_ictx->copyup_list_lock.Lock();
+    ldout(m_ictx->cct, 20) << __func__ << " removing the slot " << dendl;
+    map<uint64_t, CopyupRequest*>::iterator it =
+      m_ictx->copyup_list.find(m_object_no);
+    assert(it != m_ictx->copyup_list.end());
+    it->second = NULL;
+    m_ictx->copyup_list.erase(it);
+
+    if (m_ictx->copyup_list.empty())
+      m_ictx->copyup_list_cond.Signal();
+
+    ldout(m_ictx->cct, 20) <<  __func__ << " remove the slot " << m_object_no
+                           << " in copyup_list, size = " << m_ictx->copyup_list.size()
+                           << dendl;
+
+    m_ictx->copyup_list_lock.Unlock();
+  }
+
+  void CopyupRequest::set_ready() {
+    m_ready = true;
+  }
+
+  bool CopyupRequest::is_ready() {
+    return m_ready;
+  }
+
+  bool CopyupRequest::is_need_send_copyup() {
+    return m_need_copyup;
+  }
+
+  ceph::bufferlist& CopyupRequest::get_copyup_data() {
+    return m_copyup_data;
+  }
+
+  Mutex& CopyupRequest::get_lock() {
+    return m_lock;
+  }
+
+  void CopyupRequest::append_request(AioRequest *req) {
+    assert(!m_ready);
+    m_pending_requests.push_back(req);
+  }
+
+  void CopyupRequest::complete_all(int r) {
+    assert(m_ready);
+
+    while (!m_pending_requests.empty()) {
+      vector<AioRequest *>::iterator it = m_pending_requests.begin();
+      AioRequest *req = *it;
+      req->complete(r);
+      m_pending_requests.erase(it);
+    }
+  }
+
+  void CopyupRequest::send_copyup(int r) {
+    ldout(m_ictx->cct, 20) << __func__ << dendl;
+
+    m_ictx->snap_lock.get_read();
+    ::SnapContext snapc = m_ictx->snapc;
+    m_ictx->snap_lock.put_read();
+
+    std::vector<librados::snap_t> snaps;
```

Comment:
Pass by reference to make a copy -- or just make it prettier like: ```snaps.insert(snaps.end(), snapc.snaps.begin(), snapc.snaps.end());```.  Very minor comment.